File: frontend/src/fetch/application-info.ts
import { requestJson } from "@/fetch";
import type {
	AuthenticationProtocolValue,
	ConsolidatorUsedValue,
	CurrentMfaOptionValue,
	CurrentSignInOptionValue,
	IdentityProofingMethodValue,
	PersonalInformationValue,
	UserTypeValue,
} from "@/features/workspaces/application-info-options";

export type ApplicationInfoRead = {
	aboutApplication: string;
	applicationDescription: string;
	applicationName: string;
	applicationUrl: string;
	authorityToCollectPersonalInformation: string;
	authenticationProtocol: AuthenticationProtocolValue;
	consolidatorUsed?: ConsolidatorUsedValue | null;
	createdAt: string;
	createdBy: number | null;
	credentialAssuranceLevel: string;
	currentMfaOptions?: CurrentMfaOptionValue | null;
	currentSignInOptions: Array<CurrentSignInOptionValue>;
	currentSignInOptionsOther?: string | null;
	hasAccessManagementLayer: boolean;
	hasAccountRecovery: boolean;
	hasPrivacyNotice: boolean;
	id: number;
	identityAssuranceLevel: string;
	identityProofingMethod: IdentityProofingMethodValue;
	identityProofingMethodOther?: string | null;
	isCbas: boolean;
	isDeleted: boolean;
	migrationRationale?: string | null;
	monthlyActiveUsers?: number | null;
	peakUsagePeriods?: string | null;
	personalInformationCollected: Array<PersonalInformationValue>;
	personalInformationOther?: string | null;
	plannedOidcImplementationDate?: string | null;
	portalName?: string | null;
	programLineOfBusiness: string;
	rpApplicationUuid?: string | null;
	requestsProfileDataPushes: boolean;
	rollbackStrategy: string;
	scheduleBlackoutPeriods?: string | null;
	techStack: string;
	technology: string;
	transitionMitigations?: string | null;
	transitionRisks?: string | null;
	userTypeOther?: string | null;
	userTypes: Array<UserTypeValue>;
	usesCanadaloginMigration: boolean;
	uuid: string;
	workspaceId: number;
};

export type ApplicationInfoCreate = {
	aboutApplication: string;
	applicationDescription: string;
	applicationName: string;
	applicationUrl: string;
	authorityToCollectPersonalInformation: string;
	authenticationProtocol: AuthenticationProtocolValue;
	credentialAssuranceLevel: string;
	currentSignInOptions: Array<CurrentSignInOptionValue>;
	hasAccessManagementLayer: boolean;
	hasAccountRecovery: boolean;
	hasPrivacyNotice: boolean;
	identityAssuranceLevel: string;
	identityProofingMethod: IdentityProofingMethodValue;
	isCbas: boolean;
	personalInformationCollected: Array<PersonalInformationValue>;
	programLineOfBusiness: string;
	requestsProfileDataPushes: boolean;
	rollbackStrategy: string;
	techStack: string;
	technology: string;
	userTypes: Array<UserTypeValue>;
	usesCanadaloginMigration: boolean;
	consolidatorUsed?: ConsolidatorUsedValue;
	currentMfaOptions?: CurrentMfaOptionValue;
	currentSignInOptionsOther?: string;
	identityProofingMethodOther?: string;
	migrationRationale?: string;
	monthlyActiveUsers?: number;
	peakUsagePeriods?: string;
	personalInformationOther?: string;
	plannedOidcImplementationDate?: string;
	portalName?: string;
	scheduleBlackoutPeriods?: string;
	transitionMitigations?: string;
	transitionRisks?: string;
	userTypeOther?: string;
};

export type ApplicationInfoUpdate = Partial<ApplicationInfoCreate>;

export type ApplicationContactRead = {
	action?: string | null;
	alternatePhoneNumber?: string | null;
	applicationInfoId: number;
	contactRoles: Array<string>;
	contactType?: string | null;
	createdAt: string;
	email: string;
	firstName: string;
	id: number;
	isDeleted: boolean;
	lastName: string;
	phoneNumber: string;
	titleRole: string;
	uuid: string;
};

export type ApplicationContactCreate = {
	email: string;
	firstName: string;
	lastName: string;
	phoneNumber: string;
	titleRole: string;
	contactRoles: Array<string>;
	action?: string;
	alternatePhoneNumber?: string;
	contactType?: string;
};

export type ApplicationContactUpdate = Partial<ApplicationContactCreate>;

export const workspaceApplicationInfoQueryKey = (
	workspaceUuid: string
) => ["workspace-application-info", workspaceUuid] as const;

export const applicationInfoContactsQueryKey = (
	workspaceUuid: string,
	applicationInfoUuid: string
) =>
	[
		"application-info-contacts",
		workspaceUuid,
		applicationInfoUuid,
	] as const;

export const getWorkspaceApplicationInfos = async (
	workspaceUuid: string
): Promise<Array<ApplicationInfoRead>> => {
	const result = await requestJson<Array<ApplicationInfoRead> | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info`,
		{
			cache: "no-store",
			method: "GET",
		}
	);

	return result ?? [];
};

export const createWorkspaceApplicationInfo = async (
	workspaceUuid: string,
	payload: ApplicationInfoCreate
): Promise<ApplicationInfoRead> => {
	const result = await requestJson<ApplicationInfoRead | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info`,
		{
			body: JSON.stringify(payload),
			method: "POST",
		}
	);

	if (!result) {
		throw new Error("Failed to create application information");
	}

	return result;
};

export const updateWorkspaceApplicationInfo = async (
	workspaceUuid: string,
	applicationInfoUuid: string,
	payload: ApplicationInfoUpdate
): Promise<ApplicationInfoRead> => {
	const result = await requestJson<ApplicationInfoRead | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}`,
		{
			body: JSON.stringify(payload),
			method: "PATCH",
		}
	);

	if (!result) {
		throw new Error("Failed to update application information");
	}

	return result;
};

export const deleteWorkspaceApplicationInfo = async (
	workspaceUuid: string,
	applicationInfoUuid: string
): Promise<{ message: string }> => {
	const result = await requestJson<{ message: string } | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}`,
		{
			method: "DELETE",
		}
	);

	if (!result) {
		throw new Error("Failed to delete application information");
	}

	return result;
};

export const getWorkspaceApplicationContacts = async (
	workspaceUuid: string,
	applicationInfoUuid: string
): Promise<Array<ApplicationContactRead>> => {
	const result = await requestJson<Array<ApplicationContactRead> | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}/contacts`,
		{
			cache: "no-store",
			method: "GET",
		}
	);

	return result ?? [];
};

export const createWorkspaceApplicationContact = async (
	workspaceUuid: string,
	applicationInfoUuid: string,
	payload: ApplicationContactCreate
): Promise<ApplicationContactRead> => {
	const result = await requestJson<ApplicationContactRead | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}/contacts`,
		{
			body: JSON.stringify(payload),
			method: "POST",
		}
	);

	if (!result) {
		throw new Error("Failed to create application contact");
	}

	return result;
};

export const updateWorkspaceApplicationContact = async (
	workspaceUuid: string,
	applicationInfoUuid: string,
	applicationContactUuid: string,
	payload: ApplicationContactUpdate
): Promise<ApplicationContactRead> => {
	const result = await requestJson<ApplicationContactRead | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}/contacts/${encodeURIComponent(applicationContactUuid)}`,
		{
			body: JSON.stringify(payload),
			method: "PATCH",
		}
	);

	if (!result) {
		throw new Error("Failed to update application contact");
	}

	return result;
};

export const deleteWorkspaceApplicationContact = async (
	workspaceUuid: string,
	applicationInfoUuid: string,
	applicationContactUuid: string
): Promise<{ message: string }> => {
	const result = await requestJson<{ message: string } | null>(
		`/api/v1/workspaces/${encodeURIComponent(workspaceUuid)}/application-info/${encodeURIComponent(applicationInfoUuid)}/contacts/${encodeURIComponent(applicationContactUuid)}`,
		{
			method: "DELETE",
		}
	);

	if (!result) {
		throw new Error("Failed to delete application contact");
	}

	return result;
};

File: demo/backend/src/app/legacy/backend_models/application_info.py
import uuid as uuid_pkg
from datetime import UTC, datetime

from sqlalchemy import JSON, Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from uuid6 import uuid7

from ..core.db.database import Base


class ApplicationInfo(Base):
    __tablename__ = "application_info"

    id: Mapped[int] = mapped_column(autoincrement=True, primary_key=True, init=False)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("workspace.id"), index=True, nullable=False)
    application_name: Mapped[str] = mapped_column(String(255), nullable=False)
    about_application: Mapped[str] = mapped_column(Text, nullable=False)
    program_line_of_business: Mapped[str] = mapped_column(String(255), nullable=False)
    application_url: Mapped[str] = mapped_column(String(500), nullable=False)
    application_description: Mapped[str] = mapped_column(Text, nullable=False)
    technology: Mapped[str] = mapped_column(String(255), nullable=False)
    authentication_protocol: Mapped[str] = mapped_column(String(100), nullable=False)
    tech_stack: Mapped[str] = mapped_column(Text, nullable=False)
    rollback_strategy: Mapped[str] = mapped_column(Text, nullable=False)
    credential_assurance_level: Mapped[str] = mapped_column(String(100), nullable=False)
    identity_assurance_level: Mapped[str] = mapped_column(String(100), nullable=False)
    identity_proofing_method: Mapped[str] = mapped_column(String(255), nullable=False)
    authority_to_collect_personal_information: Mapped[str] = mapped_column(Text, nullable=False)
    uuid: Mapped[uuid_pkg.UUID] = mapped_column(UUID(as_uuid=True), default_factory=uuid7, unique=True)
    portal_name: Mapped[str | None] = mapped_column(String(255), default=None, nullable=True)
    planned_oidc_implementation_date: Mapped[str | None] = mapped_column(String(100), default=None, nullable=True)
    requests_profile_data_pushes: Mapped[bool] = mapped_column(Boolean, default=False)
    has_access_management_layer: Mapped[bool] = mapped_column(Boolean, default=False)
    identity_proofing_method_other: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    is_cbas: Mapped[bool] = mapped_column(Boolean, default=False)
    has_account_recovery: Mapped[bool] = mapped_column(Boolean, default=False)
    has_privacy_notice: Mapped[bool] = mapped_column(Boolean, default=False)
    user_types: Mapped[list[str]] = mapped_column(JSON, default_factory=list)
    user_type_other: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    monthly_active_users: Mapped[int | None] = mapped_column(Integer, default=None, nullable=True)
    peak_usage_periods: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    personal_information_collected: Mapped[list[str]] = mapped_column(JSON, default_factory=list)
    personal_information_other: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    current_sign_in_options: Mapped[list[str]] = mapped_column(JSON, default_factory=list)
    current_sign_in_options_other: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    consolidator_used: Mapped[str | None] = mapped_column(String(255), default=None, nullable=True)
    current_mfa_options: Mapped[str | None] = mapped_column(String(255), default=None, nullable=True)
    uses_canadalogin_migration: Mapped[bool] = mapped_column(Boolean, default=False)
    migration_rationale: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    schedule_blackout_periods: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    transition_risks: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    transition_mitigations: Mapped[str | None] = mapped_column(Text, default=None, nullable=True)
    created_by: Mapped[int | None] = mapped_column(ForeignKey("user.id"), default=None, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default_factory=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), default=None)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), default=None)
    is_deleted: Mapped[bool] = mapped_column(default=False, index=True)
